arch/x86/pagetables.c

```diff
@@ -446,6 +446,103 @@ int vunmap_user(void *va, mfn_t *mfn, unsigned int *order) {
     return err;
 }
 
+static int get_va_mfn_order(const cr3_t *cr3_ptr, const void *va, mfn_t *mfn,
+                            unsigned int *order) {
+    unsigned int _order;
+    mfn_t _mfn;
+    pgentry_t *tab;
+
+    ASSERT(mfn || order);
+    if (mfn_invalid(cr3_ptr->mfn))
+        return -EINVAL;
+
+    tab = tmp_map_mfn(cr3_ptr->mfn);
+#if defined(__x86_64__)
+    pml4_t *l4e = l4_table_entry((pml4_t *) tab, va);
+    if (mfn_invalid(l4e->mfn) || !l4e->P)
+        return -ENOENT;
+
+    tab = tmp_map_mfn(l4e->mfn);
+#endif
+    pdpe_t *l3e = l3_table_entry((pdpe_t *) tab, va);
+    if (l3e->PS) {
+        _mfn = l3e->mfn;
+        _order = PAGE_ORDER_1G;
+        goto done;
+    }
+
+    if (mfn_invalid(l3e->mfn) || !l3e->P)
+        return -ENOENT;
+
+    tab = tmp_map_mfn(l3e->mfn);
+    pde_t *l2e = l2_table_entry((pde_t *) tab, va);
+    if (l2e->PS) {
+        _mfn = l2e->mfn;
+        _order = PAGE_ORDER_2M;
+        goto done;
+    }
+
+    if (mfn_invalid(l2e->mfn) || !l2e->P)
+        return -ENOENT;
```

Comment:
Same here as well. If the present bit isn't set, the other bits have no meaning, according to the SDM.